### PR TITLE
TV Event Handler

### PR DESCRIPTION
**Pull request by chatrapathi06**
1. Added the EventKeyAction and Event Type to onKey method to the respective platforms.
2. created the Platform-specific mapping (X11 and WPE) functions.  
3. Conver the RNS keys to string and sent to the application.

### CODE REVIEW

File: ReactSkia/BUILD.gn [hunk 1]
```diff
@@ -60,6 +61,8 @@ source_set("ReactSkia") {
     "RNInstance.h",
     "sdk/NotificationCenter.cpp",
     "sdk/NotificationCenter.h",
+    "sdk/TVNavigationEventEmitter.cpp",
```
Comment (hps1978) [inline ReactSkia/BUILD.gn]: The filenames should be RSkTVNavigationEventEmitter.cpp and RSkTVNavigationEventEmitter.h and should be under "ReactSkia/modules" folder

Also update the class name to reflect the same name in those files

File: ReactSkia/JSITurboModuleManager.cpp [hunk 2]
```diff
@@ -226,7 +226,11 @@ JSITurboModuleManager::JSITurboModuleManager(Instance *bridgeInstance)
 
   modules_["Timing"] =
       std::make_shared<TimingModule>("Timing", jsInvoker, bridgeInstance);
-
+  std::cout <<"Mapping a event handler"<<std::endl;
```
Comment (hps1978) [inline ReactSkia/JSITurboModuleManager.cpp]: Mapping TV Events handler

File: ReactSkia/RNInstance.cpp [hunk 3]
```diff
@@ -36,6 +36,9 @@
 
 namespace facebook {
 namespace react {
+void reactGlogLoggingHook(const std::string &message,unsigned int logLevel) {
```
Comment (hps1978) [inline ReactSkia/RNInstance.cpp]: not sure what these changes are for in this file and how are they related to TV Event Handler?

File: ReactSkia/core_modules/RSkEventEmitter.h [hunk 4]
```diff
@@ -52,4 +52,4 @@ class RSkEventEmitter: public TurboModule {
       int listenerCount_;
 };
 } // namespace react
-} // namespace facebook
\ No newline at end of file
+} // namespace facebook
```
Comment (hps1978) [inline ReactSkia/core_modules/RSkEventEmitter.h:55]: why is this change required?

File: ReactSkia/sdk/RNSKeyCodeMapping.h [hunk 5]
```diff
@@ -0,0 +1,94 @@
+#ifndef __RNSKeyCodeMapping_h__
```
Comment (hps1978) [inline ReactSkia/sdk/RNSKeyCodeMapping.h]: should this be defined as __RNS_KEYCODE_MAPPING_H__?
Please confirm and use the standard used across RN Skia codebase

File: ReactSkia/sdk/NotificationCenter.h [hunk 6]
```diff
@@ -28,7 +28,7 @@
 #include <mutex>
 #include <list>
 #include <algorithm>
-
+#include <iostream>
```
Comment (hps1978) [inline ReactSkia/sdk/NotificationCenter.h]: Use the standard logging APIs to print logs

File: ReactSkia/sdk/NotificationCenter.h [hunk 7]
```diff
@@ -108,9 +108,10 @@ unsigned int NotificationCenter::on(std::string eventName, std::function<void (A
 template <typename... Args>
 void NotificationCenter::emit(std::string eventName, Args... args) {
     std::list<std::shared_ptr<Listener<Args...>>> handlers;
-    
     {
-        std::lock_guard<std::mutex> lock(mutex);
+	std::cout <<"[NotificationCenter::emit] entry"<<std::endl;
```
Comment (hps1978) [inline ReactSkia/sdk/NotificationCenter.h]: use the standard logging APIs to print logs

File: ReactSkia/sdk/RNSKeyCodeMapping.h [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+#define __RNSKeyCodeMapping_h__
+#include <iostream>
```
Comment (hps1978) [inline ReactSkia/sdk/RNSKeyCodeMapping.h:3]: why is iostream required in this .h file?

File: ReactSkia/sdk/TVNavigationEventEmitter.cpp [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+RSkTVNavigationEventEmitter::RSkTVNavigationEventEmitter(
+    const std::string &name, 
+    std::shared_ptr<CallInvoker> jsInvoker,
+    Instance *bridgeInstance)
+    : RSkEventEmitter(name, jsInvoker, bridgeInstance)
+{
+
+}
+void RSkTVNavigationEventEmitter::onTVKeyEvent(int  eventType,int eventKeyAction/*, const unsigned int tag*/) {
+    cout <<"RSkTVNavigationEventEmitter::onTVKeyEvent "<<endl;
```
Comment (hps1978) [inline ReactSkia/sdk/TVNavigationEventEmitter.cpp]: remove iostream and use the standard logging APIs throughout this file

File: ReactSkia/sdk/TVNavigationEventEmitter.cpp [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+    folly::dynamic parameters = folly::dynamic::object();
+    parameters["eventType"] = RNSKeyMap[eventType];
+    parameters["tag"] = 0xFFFF;
```
Comment (hps1978) [inline ReactSkia/sdk/TVNavigationEventEmitter.cpp]: could you please confirm the default tag value (as per our tests on Android)? 0xFFFF was just set as a default for testing only.

File: ReactSkia/sdk/TVNavigationEventEmitter.h [hunk 11]
```diff
@@ -0,0 +1,37 @@
+#include "NotificationCenter.h"
+#include "cxxreact/Instance.h"
```
Comment (hps1978) [inline ReactSkia/sdk/TVNavigationEventEmitter.h]: this does not needs to be included as the RSkEventEmitter.h already does that

File: ReactSkia/modules/RSkTVNavigationEventEmitter.h [hunk 12]
```diff
@@ -0,0 +1,36 @@
+#include "sdk/NotificationCenter.h"
```
Comment (pprabhu-ng) [inline ReactSkia/modules/RSkTVNavigationEventEmitter.h:1]: Can we have all these header files absolute paths from ReactSkia.,Say - ReactSkia/sdk/NotificationCenter.h
Reasons :
1. Following the ReactSkia coding format in UI components  , to maintain uniformity
2. It will help to know whether this files are from ReactSkia or ReactCommon

File: ReactSkia/sdk/RNSKeyCodeMapping.h [hunk 13]
```diff
@@ -0,0 +1,94 @@
+#ifndef RNS_KEYCODE_MAPPING_H
+#define RNS_KEYCODE_MAPPING_H
+typedef enum _rns_key
```
Comment (pprabhu-ng) [inline ReactSkia/sdk/RNSKeyCodeMapping.h]: enum name need to be camelcase format

File: rns_shell/BUILD.gn [hunk 14]
```diff
@@ -10,7 +10,9 @@ config("rns_shell_config") {
   include_dirs = [ "." ]
 
   if(is_linux) {
-    include_dirs += [ "." ]
+    include_dirs += [ ".",
+                      "//ReactSkia/sdk/",
+                      "//sk_app/unix/platform/window/x11" ]
```
Comment (pprabhu-ng) [inline rns_shell/BUILD.gn]: why is sk_app required here ? 
sk_app and RNS_SHELL are mutually exclusive and sk_app would be deleted soon.
So including any files from sk_app in RNS_Shell need to be avoided.

File: rns_shell/platform/graphics/libwpe/WindowLibWPE.cpp [hunk 15]
```diff
@@ -51,8 +51,14 @@ bool WindowLibWPE::initViewBackend(wpe_view_backend* viewBackend) {
     static struct wpe_view_backend_input_client s_inputClient = {
         // handle_keyboard_event
         [](void* data, struct wpe_input_keyboard_event* event) {
-            RNS_LOG_NOT_IMPL;
+            //RNS_LOG_NOT_IMPL;
```
Comment (pprabhu-ng) [inline rns_shell/platform/graphics/libwpe/WindowLibWPE.cpp]: If implemented, then this line can be deleted instead of commenting

File: rns_shell/platform/graphics/libwpe/WindowLibWPE.cpp [hunk 16]
```diff
@@ -51,8 +51,14 @@ bool WindowLibWPE::initViewBackend(wpe_view_backend* viewBackend) {
...
             auto& winwpe = *reinterpret_cast<WindowLibWPE*>(data);
+            int keycode = winwpe.keyIdentifierForWPEKeyCode(event->key_code);
+            if(event->pressed){
```
Comment (pprabhu-ng) [inline rns_shell/platform/graphics/libwpe/WindowLibWPE.cpp]: Why is this if condition required here ?
Both if and else cases seem to be doing the same

File: rns_shell/platform/graphics/libwpe/WindowLibWPE.cpp [hunk 17]
```diff
@@ -10,9 +10,9 @@
 #include "src/utils/SkUTF.h"
 #include "WindowContextFactory.h"
 #include "WindowLibWPE.h"
-
 #include "platform/linux/TaskLoop.h"
-
+#include "RNSKeyCodeMapping.h"
```
Comment (pprabhu-ng) [inline rns_shell/platform/graphics/libwpe/WindowLibWPE.cpp]: This include is probably not required. Since WindowLibWPE.h is already including it.

File: rns_shell/platform/graphics/x11/WindowX11.cpp [hunk 18]
```diff
@@ -21,7 +21,15 @@
 #include "WindowContextFactory.h"
 
 #include "x11/PlatformDisplayX11.h"
-
+//#include "tools/skui/ModifierKey.h"
```
Comment (pprabhu-ng) [inline rns_shell/platform/graphics/x11/WindowX11.cpp]: cleanup of these commented lines if not required .

File: rns_shell/platform/graphics/x11/WindowX11.cpp [hunk 19]
```diff
@@ -21,7 +21,15 @@
...
+//#include "tools/timer/Timer.h"
+#include <X11/X.h>
```
Comment (pprabhu-ng) [inline rns_shell/platform/graphics/x11/WindowX11.cpp]: <X11/X.h> is already included in WindowX11.h
Whether this include is required again ?

File: rns_shell/platform/graphics/x11/WindowX11.cpp [hunk 20]
```diff
@@ -231,6 +239,13 @@ void WindowX11::closeWindow() {
 }
 
 bool WindowX11::handleEvent(const XEvent& event) {
+      int keycode =1;
```
Comment (pprabhu-ng) [inline rns_shell/platform/graphics/x11/WindowX11.cpp]: can this be initialized with -1 value , as 1 is valid keycode.

File: rns_shell/platform/graphics/libwpe/WindowLibWPE.h [hunk 21]
```diff
@@ -14,8 +14,10 @@
 #include "src/core/SkTDynamicHash.h"
 
 #include "Window.h"
+#include "NotificationCenter.h"
 #include "PlatformDisplay.h"
 #include "platform/graphics/libwpe/PlatformDisplayLibWPE.h"
+#include "RNSKeyCodeMapping.h"
```
Comment (pprabhu-ng) [inline rns_shell/platform/graphics/libwpe/WindowLibWPE.h]: Can we have absolute paths for these files ?
Looking at the include statement itself , it will help us to know this file is included from which layer - ReactSkia or ReactCommon or RNS_Shell especially if its inter-layer header file includes.
@munezbn  what do u suggest ?

File: ReactSkia/modules/RSkTVNavigationEventEmitter.cpp [hunk 22]
```diff
@@ -0,0 +1,35 @@
+#include "RSkTVNavigationEventEmitter.h"
+#include <glog/logging.h>
```
Comment (pprabhu-ng) [inline ReactSkia/modules/RSkTVNavigationEventEmitter.cpp]: glog/logging.h usage to be replaced with RNSLog.h

File: ReactSkia/sdk/TVNavigationEventEmitter.cpp (hunk 10 above)
Comment (chatrapathi06) [reply ReactSkia/sdk/TVNavigationEventEmitter.cpp]: Sure.

File: rns_shell/platform/graphics/libwpe/WindowLibWPE.h (hunk 21 above)
Comment (chatrapathi06) [reply rns_shell/platform/graphics/libwpe/WindowLibWPE.h]: Added the absolute path

File: ReactSkia/modules/RSkTVNavigationEventEmitter.h [hunk 23]
```diff
@@ -0,0 +1,36 @@
...
+#include "tools/skui/Key.h"
+
+namespace facebook {
+namespace react {
+
+class RSkTVNavigationEventEmitter final: public RSkEventEmitter {
+    private:
+        std::vector<std::string> events_ = {"onHWKeyEvent"};
+        std::string tvEventName_ = "RCTTVNavigationEventNotification";
+        unsigned int eventId_ = 0xFFF; // Some undefined ID
```
Comment (munezbn) [inline ReactSkia/modules/RSkTVNavigationEventEmitter.h]: I would suggest use max UINT value here

File: packages/react-native-skia/modules/tvEventHandler/SimpleViewApp.js [hunk 24]
```diff
@@ -0,0 +1,77 @@
...
+      <Text style={{ color: '#fff', fontSize: 36,
+                    textAlign: 'center', marginTop: 32 }}>
+        React Native loves Skia
+      </Text>
+    </View>
+    );
+  }
+}
+
+AppRegistry.registerComponent('SimpleViewApp', () => SimpleViewApp);
```
Comment (munezbn) [inline packages/react-native-skia/modules/tvEventHandler/SimpleViewApp.js]: Can we rename this app. something like TvEventTestApp

File: rns_shell/BUILD.gn [hunk 25]
```diff
@@ -10,7 +10,8 @@ config("rns_shell_config") {
   include_dirs = [ "." ]
 
   if(is_linux) {
-    include_dirs += [ "." ]
+    include_dirs += [ ".",
+                      "//ReactSkia/sdk/" ]
```
Comment (munezbn) [inline rns_shell/BUILD.gn]: This is not specific to linux.  You can add it before if linux check

File: rns_shell/BUILD.gn [hunk 26]
```diff
@@ -48,6 +49,7 @@ if(is_linux && gl_display_backend == "x11") {
     "platform/graphics/x11/PlatformDisplayX11.cpp",
     "platform/graphics/x11/WindowX11.h",
     "platform/graphics/x11/WindowX11.cpp",
+    "platform/graphics/x11/X11KeyMap.cpp",
```
Comment (munezbn) [inline rns_shell/BUILD.gn:51]: The file name is different compared to WPE. Better to keep Uniform names. X11KeyCodeMapping.cpp

File: rns_shell/platform/graphics/x11/X11KeyMap.cpp [hunk 27]
```diff
@@ -0,0 +1,32 @@
...
+    case XK_2: return RNS_KEY_2;
+    case XK_3: return RNS_KEY_3;
+    case XK_4: return RNS_KEY_4;
+    case XK_5: return RNS_KEY_5;
+    case XK_6: return RNS_KEY_6;
+    case XK_7: return RNS_KEY_7;
+    case XK_8: return RNS_KEY_8;
+    case XK_9: return RNS_KEY_9;
+    case XK_Cancel: return RNS_KEY_Cancel;
+    default: return RNS_KEY_UnKnown;
```
Comment (munezbn) [inline rns_shell/platform/graphics/x11/X11KeyMap.cpp:28]: Why only limited keys defined here compared to WPE ?

File: ReactSkia/modules/RSkTVNavigationEventEmitter.h [hunk 28]
```diff
@@ -0,0 +1,36 @@
...
+
+        NotificationCenter keyEventNotification;
+
+        void onTVKeyEvent(int  eventType,int eventKeyAction/*, const unsigned int*/);
```
Comment (munezbn) [inline ReactSkia/modules/RSkTVNavigationEventEmitter.h]: add tag as default value here. and modify the function accordingly.

File: rns_shell/platform/graphics/x11/WindowX11.cpp [hunk 29]
```diff
@@ -21,7 +21,9 @@
 #include "WindowContextFactory.h"
 
 #include "x11/PlatformDisplayX11.h"
-
+#define X11KeyRelease 0
+#define X11KeyPress   1
```
Comment (munezbn) [inline rns_shell/platform/graphics/x11/WindowX11.cpp]: I think we need to define RNS_KEY_Press and Release enum for this and use in both X11 & WPE backend

**Review by munezbn — CHANGES_REQUESTED** [review]

File: react-native/Libraries/Components/AppleTV/TVEventHandler.js [hunk 30]
```diff
@@ -0,0 +1,55 @@
...
+  __nativeTVNavigationEventEmitter: ?NativeEventEmitter = null;
+
+  enable(component: ?any, callback: Function): void {
+    if (Platform.OS === 'ios' && !NativeTVNavigationEventEmitter) {
+      return;
+    }
+    
+    console.log('Before calling NativeEventEmitter: ', NativeEventEmitter, NativeTVNavigationEventEmitter);
+    this.__nativeTVNavigationEventEmitter = new NativeEventEmitter(
+      NativeTVNavigationEventEmitter,
```
Comment (pprabhu-ng) [inline react-native/Libraries/Components/AppleTV/TVEventHandler.js]: This file is not required to be committed

File: react-native/index.js [hunk 31]
```diff
@@ -0,0 +1,660 @@
+/**
```
Comment (pprabhu-ng) [inline react-native/index.js]: This file can be removed

**Review by pprabhu-ng — APPROVED** [review]

**Review by munezbn — APPROVED** [review]
